### PR TITLE
Bug 1517303 - Add tests to verify CFR & targeting

### PR DESCRIPTION
**Pull request by rlr**
@k88hudson r? was this what you had in mind or was there more? These tests assume that the trigger is setting up the context properly.

### CODE REVIEW

File: test/browser/browser_asrouter_targeting.js [hunk 1]
```diff
@@ -434,3 +435,41 @@ add_task(async function check_pinned_tabs() {
...
+
+  trigger.context.recentVisits.push({timestamp: timeMinutesAgo(59)});
+  is(await ASRouterTargeting.findMatchingMessage({messages, trigger}), messages.find(m => m.id === "PIN_TAB"),
+    "should select PIN_TAB mesage");
+
+  await BrowserTestUtils.withNewTab({gBrowser, url: "about:blank"}, async browser => {
+    let tab = gBrowser.getTabForBrowser(browser);
+    gBrowser.pinTab(tab);
+    is(await ASRouterTargeting.findMatchingMessage({messages, trigger}), undefined,
+      "should not select PIN_TAB mesage with only 2 visits in past hour");
```
Comment (piatra) [inline test/browser/browser_asrouter_targeting.js]: I think the message description here should be updated to mention the pinned tab?

File: test/browser/browser_asrouter_targeting.js [hunk 2]
```diff
@@ -434,3 +435,41 @@ add_task(async function check_pinned_tabs() {
...
+  });
+
+  trigger.param = "foo.bar";
+  is(await ASRouterTargeting.findMatchingMessage({messages, trigger}), undefined,
+    "should not select PIN_TAB mesage with a param/host not in our list");
```
Comment (piatra) [inline test/browser/browser_asrouter_targeting.js]: Also here the trigger is not related to the hostlist.

File: test/browser/browser_asrouter_targeting.js [hunk 3]
```diff
@@ -434,3 +435,41 @@ add_task(async function check_pinned_tabs() {
...
+
+  // Cleanup
+  await clearHistoryAndBookmarks();
```
Comment (piatra) [inline test/browser/browser_asrouter_targeting.js]: Is this required? I think you're only adding visits in the local `trigger` object.

File: test/browser/browser_asrouter_targeting.js [hunk 4]
```diff
@@ -434,3 +435,41 @@ add_task(async function check_pinned_tabs() {
...
+  is(await ASRouterTargeting.findMatchingMessage({messages, trigger}), undefined,
+    "should not select PIN_TAB mesage with only 2 visits in past hour");
+
+  trigger.context.recentVisits.push({timestamp: timeMinutesAgo(59)});
+  is(await ASRouterTargeting.findMatchingMessage({messages, trigger}), messages.find(m => m.id === "PIN_TAB"),
+    "should select PIN_TAB mesage");
+
+  await BrowserTestUtils.withNewTab({gBrowser, url: "about:blank"}, async browser => {
+    let tab = gBrowser.getTabForBrowser(browser);
+    gBrowser.pinTab(tab);
```
Comment (piatra) [inline test/browser/browser_asrouter_targeting.js]: You should probably unpin at the end of the test or do it a `registerCleanupFunction`.

File: test/browser/browser_asrouter_targeting.js (hunk 1 above)
Comment (rlr) [reply test/browser/browser_asrouter_targeting.js]: doh. lots of copy/paste fails here. fixed.

**Pull request merged**